File: frontend/src/services/conversations.ts
import { supabase } from "@/lib/supabase/client";

const API_URL =
  process.env.NEXT_PUBLIC_API_URL ||
  "http://127.0.0.1:8000";

export interface Conversation {
  id: string;
  property_id: string;
  buyer_id: string;
  seller_id: string;
  status: string;
  mode: "ai_active" | "human_active" | "closed";
  created_at: string;
  updated_at: string;
  property: {
    id: string;
    title: string;
    price: number;
    city: string;
    locality?: string;
    image: string;
    seller_id: string;
  };
  buyer: {
    id: string;
    full_name: string;
    email: string;
  };
  seller: {
    id: string;
    full_name: string;
    email: string;
  };
}

export interface Message {
  id: string;
  conversation_id: string;
  sender_id: string | null;
  sender_type: "buyer" | "seller" | "ai_agent" | "system";
  message: string;
  created_at: string;
  sender?: {
    id: string;
    full_name: string;
    email: string;
    avatar_url?: string;
  };
}

async function getHeaders() {
  await supabase.auth.getUser();
  const {
    data: { session },
  } = await supabase.auth.getSession();

  const headers: Record<string, string> = {
    "Content-Type": "application/json",
  };

  if (session?.access_token) {
    headers["Authorization"] = `Bearer ${session.access_token}`;
  }

  return headers;
}

export async function getOrCreateConversation(
  propertyId: string
): Promise<{ conversation: Conversation; is_new: boolean }> {
  const headers = await getHeaders();

  if (!headers["Authorization"]) {
    throw new Error("You must be logged in to contact the seller.");
  }

  const response = await fetch(`${API_URL}/api/conversations/`, {
    method: "POST",
    headers,
    body: JSON.stringify({ property_id: propertyId }),
  });

  if (!response.ok) {
    const error = await response.json().catch(() => null);
    throw new Error(
      error?.detail || `Failed to initiate conversation: ${response.status}`
    );
  }

  const data = await response.json();
  return {
    conversation: data.conversation,
    is_new: data.is_new,
  };
}

export async function listConversations(): Promise<Conversation[]> {
  const headers = await getHeaders();

  if (!headers["Authorization"]) {
    throw new Error("You must be logged in to view conversations.");
  }

  const response = await fetch(`${API_URL}/api/conversations/`, {
    headers,
    cache: "no-store",
  });

  if (!response.ok) {
    const error = await response.json().catch(() => null);
    throw new Error(
      error?.detail || `Failed to fetch conversations: ${response.status}`
    );
  }

  const data = await response.json();
  return data.conversations;
}

export async function getConversationMessages(
  conversationId: string
): Promise<Message[]> {
  const headers = await getHeaders();

  if (!headers["Authorization"]) {
    throw new Error("You must be logged in to view messages.");
  }

  const response = await fetch(
    `${API_URL}/api/conversations/${conversationId}/messages`,
    {
      headers,
      cache: "no-store",
    }
  );

  if (!response.ok) {
    const error = await response.json().catch(() => null);
    throw new Error(
      error?.detail || `Failed to fetch messages: ${response.status}`
    );
  }

  const data = await response.json();
  return data.messages;
}

export async function sendChatMessage(
  conversationId: string,
  message: string
): Promise<Message> {
  const headers = await getHeaders();

  if (!headers["Authorization"]) {
    throw new Error("You must be logged in to send a message.");
  }

  const response = await fetch(
    `${API_URL}/api/conversations/${conversationId}/messages`,
    {
      method: "POST",
      headers,
      body: JSON.stringify({ message }),
    }
  );

  if (!response.ok) {
    const error = await response.json().catch(() => null);
    throw new Error(
      error?.detail || `Failed to send message: ${response.status}`
    );
  }

  const data = await response.json();
  return data.message;
}

export async function updateConversationMode(
  conversationId: string,
  mode: "ai_active" | "human_active" | "closed"
): Promise<Conversation> {
  const headers = await getHeaders();

  if (!headers["Authorization"]) {
    throw new Error("You must be logged in to update conversation co-pilot mode.");
  }

  const response = await fetch(`${API_URL}/api/conversations/${conversationId}`, {
    method: "PATCH",
    headers,
    body: JSON.stringify({ mode }),
  });

  if (!response.ok) {
    const error = await response.json().catch(() => null);
    throw new Error(
      error?.detail || `Failed to update conversation mode: ${response.status}`
    );
  }

  const data = await response.json();
  return data.conversation;
}

File: backend/app/api/routes/conversations.py
from fastapi import APIRouter, Depends, HTTPException, BackgroundTasks
from pydantic import BaseModel
from typing import List, Optional
from datetime import datetime

from app.database.supabase import get_supabase
from app.core.auth import get_current_user
from app.services.ai.seller_agent import run_seller_ai_agent

router = APIRouter(
    prefix="/api/conversations",
    tags=["Conversations"],
)


class ConversationCreate(BaseModel):
    property_id: str


class MessageCreate(BaseModel):
    message: str


class ConversationUpdate(BaseModel):
    mode: Optional[str] = None
    status: Optional[str] = None


@router.post("/")
def get_or_create_conversation(
    payload: ConversationCreate,
    current_user=Depends(get_current_user),
):
    try:
        supabase = get_supabase()

        # 1. Fetch property to get seller_id
        prop_res = (
            supabase
            .table("properties")
            .select("id, seller_id")
            .eq("id", payload.property_id)
            .execute()
        )

        if not prop_res.data:
            raise HTTPException(
                status_code=404,
                detail="Property not found."
            )

        property_data = prop_res.data[0]
        seller_id = property_data.get("seller_id")

        if not seller_id:
            raise HTTPException(
                status_code=400,
                detail="Property does not have an assigned seller."
            )

        # 2. Check if buyer is trying to contact themselves
        if str(current_user.id) == str(seller_id):
            raise HTTPException(
                status_code=400,
                detail="You cannot start a conversation with yourself."
            )

        # 3. Find existing conversation
        existing = (
            supabase
            .table("conversations")
            .select("*, property:properties(id, title, price, city, locality, image, seller_id), buyer:profiles!conversations_buyer_id_fkey(id, full_name, email), seller:profiles!conversations_seller_id_fkey(id, full_name, email)")
            .eq("property_id", payload.property_id)
            .eq("buyer_id", current_user.id)
            .eq("seller_id", seller_id)
            .execute()
        )

        if existing.data:
            return {
                "success": True,
                "conversation": existing.data[0],
                "is_new": False
            }

        # 4. Create new conversation
        new_conv = {
            "property_id": payload.property_id,
            "buyer_id": current_user.id,
            "seller_id": seller_id,
            "status": "active",
            "mode": "ai_active"
        }

        insert_res = (
            supabase
            .table("conversations")
            .insert(new_conv)
            .execute()
        )

        if not insert_res.data:
            raise HTTPException(
                status_code=500,
                detail="Failed to create conversation."
            )

        new_conv_id = insert_res.data[0]["id"]

        # Fetch the complete object with joins
        complete_conv = (
            supabase
            .table("conversations")
            .select("*, property:properties(id, title, price, city, locality, image, seller_id), buyer:profiles!conversations_buyer_id_fkey(id, full_name, email), seller:profiles!conversations_seller_id_fkey(id, full_name, email)")
            .eq("id", new_conv_id)
            .single()
            .execute()
        )

        return {
            "success": True,
            "conversation": complete_conv.data,
            "is_new": True
        }

    except HTTPException:
        raise
    except Exception as error:
        print("CONVERSATION CREATE ERROR:", repr(error))
        raise HTTPException(
            status_code=500,
            detail=str(error)
        )


@router.get("/")
def list_conversations(
    current_user=Depends(get_current_user),
):
    try:
        supabase = get_supabase()

        response = (
            supabase
            .table("conversations")
            .select("*, property:properties(id, title, price, city, locality, image, seller_id), buyer:profiles!conversations_buyer_id_fkey(id, full_name, email), seller:profiles!conversations_seller_id_fkey(id, full_name, email)")
            .or_(f"buyer_id.eq.{current_user.id},seller_id.eq.{current_user.id}")
            .order("updated_at", desc=True)
            .execute()
        )

        return {
            "success": True,
            "conversations": response.data
        }

    except Exception as error:
        print("LIST CONVERSATIONS ERROR:", repr(error))
        raise HTTPException(
            status_code=500,
            detail=str(error)
        )


@router.get("/{conversation_id}/messages")
def get_messages(
    conversation_id: str,
    current_user=Depends(get_current_user),
):
    try:
        supabase = get_supabase()

        # Verify participant
        conv_res = (
            supabase
            .table("conversations")
            .select("id, buyer_id, seller_id")
            .eq("id", conversation_id)
            .execute()
        )

        if not conv_res.data:
            raise HTTPException(
                status_code=404,
                detail="Conversation not found."
            )

        conv = conv_res.data[0]
        if str(current_user.id) not in [str(conv["buyer_id"]), str(conv["seller_id"])]:
            raise HTTPException(
                status_code=403,
                detail="Access denied."
            )

        # Fetch messages
        messages_res = (
            supabase
            .table("conversation_messages")
            .select("*, sender:profiles(id, full_name, email, avatar_url)")
            .eq("conversation_id", conversation_id)
            .order("created_at", desc=False)
            .execute()
        )

        return {
            "success": True,
            "messages": messages_res.data
        }

    except HTTPException:
        raise
    except Exception as error:
        print("GET MESSAGES ERROR:", repr(error))
        raise HTTPException(
            status_code=500,
            detail=str(error)
        )


@router.post("/{conversation_id}/messages")
def send_message(
    conversation_id: str,
    payload: MessageCreate,
    background_tasks: BackgroundTasks,
    current_user=Depends(get_current_user),
):
    try:
        supabase = get_supabase()

        # Verify participant and determine sender type
        conv_res = (
            supabase
            .table("conversations")
            .select("id, buyer_id, seller_id, mode")
            .eq("id", conversation_id)
            .execute()
        )

        if not conv_res.data:
            raise HTTPException(
                status_code=404,
                detail="Conversation not found."
            )

        conv = conv_res.data[0]
        sender_type = None

        if str(current_user.id) == str(conv["buyer_id"]):
            sender_type = "buyer"
        elif str(current_user.id) == str(conv["seller_id"]):
            sender_type = "seller"
        else:
            raise HTTPException(
                status_code=403,
                detail="Access denied."
            )

        # Insert message
        msg_payload = {
            "conversation_id": conversation_id,
            "sender_id": current_user.id,
            "sender_type": sender_type,
            "message": payload.message
        }

        insert_res = (
            supabase
            .table("conversation_messages")
            .insert(msg_payload)
            .execute()
        )

        if not insert_res.data:
            raise HTTPException(
                status_code=500,
                detail="Failed to send message."
            )

        # Update conversations updated_at to bring it to top
        supabase.table("conversations").update({"updated_at": datetime.now().isoformat()}).eq("id", conversation_id).execute()

        # Handle Auto Mode switch for Seller response
        if sender_type == "seller" and conv.get("mode") == "ai_active":
            supabase.table("conversations").update({"mode": "human_active"}).eq("id", conversation_id).execute()
            print(f"Conversation {conversation_id} mode set to human_active due to seller reply.")

        # Fetch complete sent message with sender profile
        msg_id = insert_res.data[0]["id"]
        complete_msg = (
            supabase
            .table("conversation_messages")
            .select("*, sender:profiles(id, full_name, email, avatar_url)")
            .eq("id", msg_id)
            .single()
            .execute()
        )

        # Trigger Autonomous Seller AI Agent if mode is ai_active and sender is buyer
        current_mode = conv.get("mode")
        if sender_type == "buyer" and current_mode == "ai_active":
            background_tasks.add_task(run_seller_ai_agent, conversation_id)
            print(f"Background task queued for Seller AI agent on conversation {conversation_id}")

        return {
            "success": True,
            "message": complete_msg.data
        }

    except HTTPException:
        raise
    except Exception as error:
        print("SEND MESSAGE ERROR:", repr(error))
        raise HTTPException(
            status_code=500,
            detail=str(error)
        )


@router.patch("/{conversation_id}")
def update_conversation(
    conversation_id: str,
    payload: ConversationUpdate,
    current_user=Depends(get_current_user),
):
    try:
        supabase = get_supabase()

        # Verify participant
        conv_res = (
            supabase
            .table("conversations")
            .select("id, buyer_id, seller_id")
            .eq("id", conversation_id)
            .execute()
        )

        if not conv_res.data:
            raise HTTPException(
                status_code=404,
                detail="Conversation not found."
            )

        conv = conv_res.data[0]
        if str(current_user.id) not in [str(conv["buyer_id"]), str(conv["seller_id"])]:
            raise HTTPException(
                status_code=403,
                detail="Access denied."
            )

        update_data = {}
        if payload.mode is not None:
            if payload.mode not in ["ai_active", "human_active", "closed"]:
                raise HTTPException(
                    status_code=400,
                    detail="Invalid conversation mode."
                )
            update_data["mode"] = payload.mode
        if payload.status is not None:
            update_data["status"] = payload.status

        if not update_data:
            raise HTTPException(
                status_code=400,
                detail="No fields to update."
            )

        # Perform update
        update_res = (
            supabase
            .table("conversations")
            .update(update_data)
            .eq("id", conversation_id)
            .execute()
        )

        return {
            "success": True,
            "conversation": update_res.data[0]
        }
    except Exception as error:
        raise HTTPException(
            status_code=500,
            detail=f"Internal server error: {str(error)}"
        )
